### ISSUE TITLE
How to check if User Stream is active before creating a new one

### ISSUE THREAD

**Issue by kayronjm** [linvi/tweetinvi#435]
Hello all,

First time posting here (on GitHub in general). I have used Tweetinvi to create a program called "Tweet Remote Control" for astrophotography purposes (to remotely control one's observatory equipment via Twitter). More about it here: http://www.lightvortexastronomy.com/tweet-remote-control.html

When my program starts, it calls a subroutine which has the following code:

`var TwitterStream = Stream.CreateUserStream();
TwitterStream.TweetCreatedByMe += (sender, args) =>`

This creates a new user stream for tweets created by the user, which works perfectly well. However, since my program runs 24/7 at a remote observatory, it is possible for the Internet connection to drop at any point in time and then return a few minutes later. Therefore, in my program, there's a timer that ticks every two minutes. It calls the following line of code:

`Auth.SetUserCredentials(ApplicationAuthenticationKeys.strConsumerKey, ApplicationAuthenticationKeys.strConsumerSecret, strAccessToken, strAccessTokenSecret);`

The timer is meant to re-authenticate with Twitter. However, I wonder, is the user stream created earlier still active when the Internet connection drops and returns a few minutes later? If not, I would need to re-create it. I can recall the subroutine that does this, but unfortunately that means creating a new user stream every two minutes. I have observed this seems to create "multiple instances" to my program running code that responds to tweets created by the user, which I certainly want to avoid. 

My question is - is there a way to check if the user stream is active prior to deciding if to create a new one? In theory the code would have to go before I define `var TwitterStream = Stream.CreateUserStream();`, to check, but I can't figure out an efficient way to do it that works. 

Of course, if the user stream remains active when the Internet connection returns (given the line of code in my two-minute re-authentication timer), then my question is null. :) Thank you in advance!

**Comment by linvi** [comment]
Hello there,

Thank you for taking the time to contact me here! Your application looks great, I look forward to use it!

Regarding your code here are few comments that I can provide.

> This creates a new user stream for tweets created by the user, which works perfectly well. However, since my program runs 24/7 at a remote observatory, it is possible for the Internet connection to drop at any point in time and then return a few minutes later. Therefore, in my program, there's a timer that ticks every two minutes. It calls the following line of code:

You do not need to do that, you can use the `StreamStopped` event.

``` c#
var stream = Stream.CreateUserStream();

stream.StreamStopped += (sender, args) =>
{
    if (args.Exception != null || args.DisconnectMessage != null)
    {
       // A timer can be required here to reduce the number of times this is executed.
       // If the internet has stopped working for a long time, it would reduce the number 
       // of requests executed on your network.
        stream.StartStream();
    }
};
```

> The timer is meant to re-authenticate with Twitter. However, I wonder, is the user stream created earlier still active when the Internet connection drops and returns a few minutes later? If not, I would need to re-create it. I can recall the subroutine that does this, but unfortunately that means creating a new user stream every two minutes. I have observed this seems to create "multiple instances" to my program running code that responds to tweets created by the user, which I certainly want to avoid.

If a stream stops because of an internet issue, the connection will be closed and the stream will no longer retrieve any tweet if you have not restarted as explained above.

> My question is - is there a way to check if the user stream is active prior to deciding if to create a new one? In theory the code would have to go before I define var TwitterStream = Stream.CreateUserStream();, to check, but I can't figure out an efficient way to do it that works.

Yes there is. You can use the `stream.StreamState` property:

``` c#
bool isStreamRunning = stream.StreamState == StreamState.Running;
```

> Of course, if the user stream remains active when the Internet connection returns (given the line of code in my two-minute re-authentication timer), then my question is null. :) 

As explained this is not the case, I would strongly suggest that you use the event.

> Thank you in advance!

Thank you for your awesome project. I like when people have such good ideas.
Good luck with it!

**Comment by kayronjm** [comment]
Thank you so much for your excellent response. It has cleared up my issue. I have now coded in the `TwitterStream.StreamStopped += (sender, args) =>` block of code and am testing it at my own observatory. I have also commented out the code that deals with the two-minute re-authentication timer. If this works out well, it'll be the perfect solution! :)

Again, thank you so much for your work on **Tweetinvi**. It's absolutely brilliant.
